src/server/pkg/storage/chunk/writer.go

```diff
@@ -18,299 +18,249 @@ import (
+				continue
 			}
-			w.annotations = nil
-			w.first = false
+			w.writeData(data[offset : i+1])
+			w.createChunk()
+			offset = i + 1
 		}
 	}
-	if a.buf.Len() > 0 {
-		w.annotations = joinAnnotations(w.annotations, a)
+	for w.numChunkBytesAnnotation+len(data[offset:]) >= w.chunkSize.max {
+		bytesLeft := w.chunkSize.max - w.numChunkBytesAnnotation
+		w.writeData(data[offset : offset+bytesLeft])
+		w.createChunk()
+		offset += bytesLeft
 	}
-	w.lastAnnotation = a
-	return nil
+	w.writeData(data[offset:])
 }
 
-func (w *worker) setupHash(a *Annotation) error {
-	// The hash state needs to be updated to include the last copied annotation(s).
-	if w.lastCopiedAnnotation != nil {
-		w.resetHash(w.lastCopiedAnnotation)
-		buf := &bytes.Buffer{}
-		for _, dr := range w.lastCopiedAnnotation.drs {
-			if err := dr.Get(buf); err != nil {
-				return err
-			}
-		}
-		for _, b := range buf.Bytes() {
-			w.hash.Roll(b)
-		}
-		w.lastAnnotation = copyAnnotation(w.lastCopiedAnnotation)
-		w.lastCopiedAnnotation = nil
-	}
-	w.resetHash(a)
-	return nil
+func (w *Writer) writeData(data []byte) {
+	lastA := w.annotations[len(w.annotations)-1]
+	lastTag := lastA.tags[len(lastA.tags)-1]
+	lastTag.SizeBytes += int64(len(data))
+	w.numChunkBytesAnnotation += len(data)
+	w.buf.Write(data)
 }
 
-// (bryce) we are occasionally creating chunks that are 1 byte because of the initialWindow.
-// this should be addressed by min/max chunks sizes.
-func (w *worker) resetHash(a *Annotation) {
-	if w.lastAnnotation != nil && a.Data != w.lastAnnotation.Data {
-		w.hash.Reset()
-		w.hash.Write(initialWindow)
-	}
+func (w *Writer) createChunk() {
+	chunk := w.buf.Bytes()
+	w.appendToChain(func(annotations []*Annotation, prevChan, nextChan chan struct{}) error {
+		return w.processChunk(chunk, annotations, prevChan, nextChan)
+	})
+	w.numChunkBytesAnnotation = 0
+	w.resetHash()
+	w.buf = &bytes.Buffer{}
 }
 
-func (w *worker) numBytesRolled() int {
-	var numBytes int
-	for _, a := range w.annotations {
-		numBytes += a.buf.Len()
+func (w *Writer) appendToChain(f func([]*Annotation, chan struct{}, chan struct{}) error, last ...bool) {
+	// (bryce) add limiter
+	annotations := w.annotations
+	w.annotations = []*Annotation{copyAnnotation(w.annotations[len(w.annotations)-1])}
+	prevChan := w.prevChan
+	var nextChan chan struct{}
+	if len(last) == 0 || !last[0] {
+		nextChan = make(chan struct{})
 	}
-	return numBytes
+	w.prevChan = nextChan
+	w.eg.Go(func() error {
+		return f(annotations, prevChan, nextChan)
+	})
+	w.stats.chunkCount++
 }
 
-func (w *worker) put(edge bool) error {
-	var chunkBytes []byte
-	for _, a := range w.annotations {
-		chunkBytes = append(chunkBytes, a.buf.Bytes()...)
+func copyAnnotation(a *Annotation) *Annotation {
+	copyA := &Annotation{Data: a.Data}
+	if a.RefDataRefs != nil {
+		copyA.RefDataRefs = a.RefDataRefs
 	}
+	if a.NextDataRef != nil {
+		copyA.NextDataRef = &DataRef{}
+	}
+	if a.tags != nil {
```

Comment:
Could this be `len(a.tags) != 0` instead of `a.tags != nil`? I think that'll give you the same semantics, except in the case where `a.tags` is a slice of size 0, in which case the line below will panic. I'm assuming the code as it is now won't ever have a `a.tags` in that state, but that could easily change so I think it makes sense to bulletproof against it.

Reply:
Sure.